drift/core/beamtransfer.py

```diff
@@ -831,190 +831,260 @@ def _generate_svdfiles(self, regen=False, skip_svd_inv=False):
 
-                    if not skip_svd_inv:
-                        ibeam = la.pinv(beam)
-                        # Find the pseudo-inverse of the beam matrix and save to disk.
-                        dset_ibsvd[fi, :, :, :nmodes] = ibeam.reshape(
-                            self.telescope.num_pol_sky, self.telescope.lmax + 1, nmodes
+                        u3, s3 = matrix_image(
+                            bft, rtol=0.0, errmsg=("SVD3 m=%i f=%i" % (mi, fi))
                         )
+                        ut3 = np.dot(u3.T.conj(), ut2)
 
-                    # Save out the singular values for each block
-                    dset_sig[fi, :nmodes] = sig
+                        nmodes = ut3.shape[0]
 
-            # Write a few useful attributes.
-            fs.attrs["baselines"] = self.telescope.baselines
-            fs.attrs["m"] = mi
-            fs.attrs["frequencies"] = self.telescope.frequencies
+                        # Skip if nmodes is zero for some reason.
+                        if nmodes == 0:
+                            continue
 
-            fs.close()
-            fm.close()
+                        # Final products
+                        ut = ut3
+                        sig = s3[:nmodes]
+                        beam = np.dot(ut3, bfr)
 
-        # If we're part of an MPI run, synchronise here.
-        mpiutil.barrier()
+                        # Save out the evecs (for transforming from the telescope frame into the SVD basis)
+                        dset_ut[fi, :nmodes] = ut * noisew[np.newaxis, :]
 
-        # Collect the spectrum into a single file.
-        self._collect_svd_spectrum()
+                        # Save out the modified beam matrix (for mapping from the sky into the SVD basis)
+                        dset_bsvd[fi, :nmodes] = beam.reshape(
+                            nmodes, self.telescope.num_pol_sky, self.telescope.lmax + 1
+                        )
+
+                        if not skip_svd_inv:
+                            # Find the pseudo-inverse of the beam matrix and save to disk.
+                            # First try la.pinv, which uses a least-squares solver.
+                            try:
+                                ibeam = la.pinv(beam)
+                            except la.LinAlgError as e:
+                                # If la.pinv fails, try la.pinv2, which is SVD-based and
+                                # more likely to succeed. If successful, add file attribute
+                                # indicating pinv2 was used for this frequency.
+                                print(
+                                    "***pinv failure: m = %d, fi = %d. Trying pinv2..."
+                                    % (mi, fi)
+                                )
+                                try:
+                                    ibeam = la.pinv2(beam)
+                                    if "inv_bsvd_from_pinv2" not in fs.attrs.keys():
```

Comment:
how important is it that we keep track which frequencies were done with the second pseudo inverse? just asking...

Reply:
For debugging - otherwise, if you want to load the matrix and look at it yourself, you won't know which frequency caused the crash